mava/networks/retention.py

```diff
@@ -0,0 +1,309 @@
+            decay_matrix = jnp.ones((batch, chunk_size, chunk_size))
+            xi = jnp.ones((batch, chunk_size, 1))
+            next_hstate = (k_proj @ v_proj) + hstate
+        else:
+            decay_matrix = self.get_decay_matrix(dones)
+            xi = self.get_xi(dones)
+            chunk_decay = self.decay_kappa ** (chunk_size // self.n_agents)
+            delta = ~jnp.any(dones[:, :: self.n_agents], axis=1)[:, jnp.newaxis, jnp.newaxis]
+            next_hstate = (
+                k_proj @ (v_proj * decay_matrix[:, -1].reshape((batch, chunk_size, 1)))
+            ) + hstate * chunk_decay * delta
+
+        # Compute the inner chunk and cross chunk
+        cross_chunk = (q_proj @ hstate) * xi
+        inner_chunk = ((q_proj @ k_proj) * decay_matrix) @ v_proj
+
+        # Compute the final retention
+        ret = inner_chunk + cross_chunk
+        return ret, next_hstate
+
+    def recurrent(
+        self, key_n: Array, query_n: Array, value_n: Array, hstate: Array
+    ) -> Tuple[Array, Array]:
+        """Recurrent representation of the retention mechanism."""
+        # Apply projection to q_proj, k_proj, v_proj
+        q_proj = query_n @ self.w_q
+        k_proj = key_n @ self.w_k
+        v_proj = value_n @ self.w_v
+
+        # Apply the retention mechanism and update the hidden state
+        updated_hstate = hstate + (k_proj.transpose(0, -1, -2) @ v_proj)
+        ret = q_proj @ updated_hstate
+
+        return ret, updated_hstate
+
+    def get_decay_matrix(self, dones: Array) -> Array:
+        """Get the decay matrix for the full sequence based on the dones and retention type."""
+        # Extract done information at the timestep level
+        timestep_dones = dones[:, :: self.n_agents]
+
+        # Compute the decay matrix and apply timestep-based masking
+        decay_matrix = self._get_default_decay_matrix(
+            timestep_dones
+        ) * self._get_decay_matrix_mask_timestep(timestep_dones)
+
+        # Repeat decay matrix across agents
+        decay_matrix = jnp.repeat(
+            jnp.repeat(decay_matrix, self.n_agents, axis=1), self.n_agents, axis=2
+        )
+
+        # Apply a causal mask over agents if full self-retention is disabled
+        if not self.full_self_retention:
+            mask_agents = jnp.tril(jnp.ones((decay_matrix.shape[1], decay_matrix.shape[1])))
+            decay_matrix = mask_agents[None, :, :] * decay_matrix
+
+        return decay_matrix
+
+    def _get_decay_matrix_mask_timestep(self, ts_dones: Array) -> Array:
+        """Generates a mask over the timesteps based on the done status of agents."""
+        # Get the shape of the input: batch size and number of timesteps
+        batch, num_ts = ts_dones.shape
+
+        # Initialize the mask
+        timestep_mask = jnp.zeros((batch, num_ts, num_ts), dtype=bool)
+        all_false = jnp.zeros((batch, num_ts, num_ts), dtype=bool)
```

Comment:
2 things
* I think these 2 comments are not need it's obvious what is happening
* Can we use the notion seen in the MAT PR for shapes so this would be `B, T = ts_dones.shape`